documentation/specs/multithreaded-msbuild.md

```diff
@@ -0,0 +1,268 @@
+
+(This diagram is simplified: the scheduler will not create a new node for every project, so in a two-project build, the scheduler will assign both projects to the same node.)
+
+```mermaid
+sequenceDiagram
+
+box Entry Process
+    participant Scheduler
+    participant Project1
+    participant Tasks
+end
+
+box Worker Process (Node 2)
+    participant Project2
+    participant Tasks2 as Tasks
+end
+
+Scheduler ->> Project1: start
+activate Project1
+Project1 ->> Tasks: Run Task X
+activate Tasks
+Tasks ->> Project1: Task X state mutations
+deactivate Tasks
+Project1 ->> Scheduler: Build Project2
+deactivate Project1
+Scheduler -->> Project2: start
+activate Project2
+Project2 ->> Tasks2: Run Task Y
+activate Tasks2
+Tasks2 ->> Project2: Task Y state mutations
+deactivate Tasks2
+Project2 -->> Scheduler: results
+deactivate Project2
+Scheduler ->> Project1: Project2 outputs
+activate Project1
+Project1 ->> Tasks: Run Task Z
+activate Tasks
+Tasks ->> Project1: Task Z state mutations
+deactivate Tasks
+Project1 ->> Scheduler: results
+deactivate Project1
+```
+
+A task that needs to run out of process can spawn a new process, as in a single-proc build, with potentially many TaskHost processes running concurrently.
+
+
+## Multithreading MSBuild
+
+The goal of multithreading MSBuild is to allow tasks to run concurrently within the same process, rather than spawning separate processes for each node. This would enable better resource utilization and potentially faster builds, as fewer processes would need to be created, reducing .NET runtime overhead and inter-process communication as well as increasing the efficacy of task-level caching.
+
+But tasks are not currently designed to be multithreaded. They assume that they own the whole process while they are executing, and they mutate the process state (environment, working directory, etc.). To make MSBuild multithreaded while maintaining task compatibility, we need to consider what parts of the build process can be moved to a multithreaded model without breaking existing tasks, as well as allowing new tasks to opt into multithreading.
+
+The scheduler is already capable of juggling multiple projects, and there's already an abstraction layer for "where a project runs" (`INodeProvider`). To support multithreading, we will introduce a new type of node, called a "thread node", which represents a thread within either an in-process or out-of-process node. This will allow us to run multiple projects concurrently within the same process.
+
+The scheduler should  be responsible for creating the appropriate combination of nodes (in-proc, out-of-proc, and thread nodes) based on the execution mode (multi-proc or multithreaded, cli or Visual Studio scenarios). It will then coordinate projects execution through the node abstraction. Below is the diagram for cli multi-threaded mode, we will create all the thread nodes in the entry process.
```

Comment:
> Below is the diagram for cli multi-threaded mode, we will create all the thread nodes in the entry process.

is this for the diagram only or we decided that all thread nodes will be in the entry node (in-proc node)?

Reply:
For CLI scenario - yes, this is current design, but I do not want to make this a limitation. First, there is VS for which we will have those in one or two out-of-process nodes, see discussion #11939. Second, I would say that we need to account for the situation that in future we might want to have multiple processes with thread nodes. So, if there is a way to make our design extendable in this direction, we probably should.